### DISCUSSION TITLE
Certificate Error

### DISCUSSION THREAD

**Discussion by jaredebird** [thingworx-field-work/ThingworxVSCodeProject#56] [Q&A] [answered]
I am getting the error below when I run npm run upload.  We are using a self signed certificate as this server is not available from outside the local network.  I'm not sure how to get it to accept the certificate and upload the file. Any suggestions?

```
✔ Created extension zip/bm-thingworx-vscode-2.5.8.zip
❯ Uploading to https://dev-thingworx.toro.com/node:internal/process/promises:288
            triggerUncaughtException(err, true /* fromPromise */);
            ^

[Error: unable to verify the first certificate] {
  code: 'UNABLE_TO_VERIFY_LEAF_SIGNATURE'
}

```

**Comment by BogdanMihaiciuc** [comment, marked as answer]
You would need to get node to trust that certificate. Ideally you would use something like the [`NODE_EXTRA_CA_CERTS`](https://nodejs.org/api/cli.html#cli_node_extra_ca_certs_file) environment variable and point to your self signed certificate. There is also the `NODE_TLS_REJECT_UNAUTHORIZED` environment variable that disables certificate verification altogether but I wouldn't recommend using it since you probably still want to verify certs on other connections.

**Reply by jaredebird** [reply]
I was able to get it to work with NODE_TLS_REJECT_UNAUTHORIZED=0 added to the top of my .env file, but I would obviously prefer not to keep using this unsecured. 

So I have tried to export the security certificate from the site.  I renamed it cert.pem and put it in the root directory of my project.  I added NODE_EXTRA_CA_CERTS=cert.pem to the first line of my .env file but I still get the same error.  

The certificate that I exported from chrome looks like this.  Is this what I need to be pointing to in my environment variable or do I need to be getting the certificate from somewhere else? 

`-----BEGIN CERTIFICATE-----
MIIHTDCCBTSgAwIBAgITLAABYITH2xipk9/gmQAAAAFghDANBgkqhkiG9w0BAQsF
The rest of the certificate removed before posting...
-----END CERTIFICATE-----
`

**Reply by BogdanMihaiciuc** [reply]
Try specifying the entire path to the cert file

**Reply by BogdanMihaiciuc** [reply]
Also the docs say this

`The NODE_EXTRA_CA_CERTS environment variable is only read when the Node.js process is first launched. Changing the value at runtime using process.env.NODE_EXTRA_CA_CERTS has no effect on the current process.`

It's likely that `dotenv` loads the env variables at runtime, so you could also trying adding that to the OS variables to ensure it's there when the node process starts.

**Reply by jaredebird** [reply]
Thank you!  Adding the cert to the OS variables is what finally got this working!  I greatly appreciate your help!

**Reply by BogdanMihaiciuc** [reply]
Excellent, glad it worked
